frontend_tests/puppeteer_tests/recent-topic.ts

```diff
@@ -0,0 +1,446 @@
+import {strict as assert} from "assert";
+
+import type {Page} from "puppeteer";
+
+import common from "../puppeteer_lib/common";
+
+const recent_topic_page = "#recent_topics_view";
+const all_filter_button = ".btn-recent-filters[data-filter='all']";
+const include_muted_filter_button = ".btn-recent-filters[data-filter='include_muted']";
+const unread_filter_button = ".btn-recent-filters[data-filter='unread']";
+const participated_filter_button = ".btn-recent-filters[data-filter='participated']";
+
+// Returns an array based on all text found within the selector passed.
+async function get_selector_text(page: Page, selector: string): Promise<string[]> {
+    const elements = await page.$$(selector);
+    const text = await Promise.all<string>(
+        elements.map(async (element) => common.get_element_text(element)),
+    ).then((value) => value);
+
+    return text;
+}
+
+async function trigger_filter_button(page: Page, selector: string): Promise<void> {
+    await page.waitForSelector(selector);
+    await page.click(selector);
+}
+
+async function get_stream_names(page: Page): Promise<string[]> {
+    await page.waitForSelector(recent_topic_page);
+
+    const stream_name_cells = ".recent_topic_stream a";
+    const stream_names = await get_selector_text(page, stream_name_cells);
+
+    return stream_names;
+}
+
+async function get_topic_names(page: Page): Promise<string[]> {
+    await page.waitForSelector(recent_topic_page);
+
+    const topic_name_cells = ".recent_topic_name a";
+    const topic_names = await get_selector_text(page, topic_name_cells);
+
+    return topic_names;
+}
+
+// Waits for the recent topics to filter once typed a certain query within
+// `filter topic` input field.
+async function wait_for_topics_to_filter(page: Page, no_of_topics: number): Promise<void> {
+    await page.waitForSelector(recent_topic_page);
+
+    await page.waitForFunction(
+        (recent_topic_page: string, no_of_topics: number) => {
+            const topics = $(recent_topic_page).find("tbody").find("tr");
+            return topics.length === no_of_topics;
+        },
+        {},
+        recent_topic_page,
+        no_of_topics,
+    );
+}
```

Comment:
This is causing a nondeterminstic failure when targeted it to run for 100x in CI. <a href = 'https://github.com/aryanshridhar/zulip/runs/2355693318'>Here's</a> the link to workflow which is failing at around 20ish iteration.
As the workflow suggestes, it's causing because of this (`wait_for_topics_to_filter`) block of function since the `waitForFunction` isn't returning a truth value and hence causing a `TimeoutError` error, 

Would love any suggestions regarding it :)